File: api/internal/tailoring/openai.go
package tailoring

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"prismapply/api/internal/config"
)

// CallJSONLLM invokes OpenAI chat completions with strict JSON schema output.
func CallJSONLLM(ctx context.Context, cfg config.Config, system, user string, schemaName string, schema map[string]any, temperature float64, dest any) error {
	apiKey := strings.TrimSpace(cfg.OpenAIAPIKey)
	if apiKey == "" {
		return fmt.Errorf("OPENAI_API_KEY is not set")
	}
	model := strings.TrimSpace(cfg.TailorModel)
	if model == "" {
		model = "gpt-4o-mini"
	}
	baseURL := strings.TrimRight(strings.TrimSpace(cfg.OpenAIBaseURL), "/")
	if baseURL == "" {
		baseURL = "https://api.openai.com/v1"
	}

	body := map[string]any{
		"model": model,
		"messages": []map[string]string{
			{"role": "system", "content": system},
			{"role": "user", "content": user},
		},
		"response_format": map[string]any{
			"type": "json_schema",
			"json_schema": map[string]any{
				"name":   schemaName,
				"strict": true,
				"schema": schema,
			},
		},
		"temperature": temperature,
	}
	raw, err := json.Marshal(body)
	if err != nil {
		return err
	}

	reqCtx, cancel := context.WithTimeout(ctx, 3*time.Minute)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, baseURL+"/chat/completions", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("openai chat error (HTTP %d): %s", resp.StatusCode, strings.TrimSpace(string(respBody)))
	}

	var parsed struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return err
	}
	if len(parsed.Choices) == 0 || strings.TrimSpace(parsed.Choices[0].Message.Content) == "" {
		return fmt.Errorf("empty LLM response")
	}
	return json.Unmarshal([]byte(parsed.Choices[0].Message.Content), dest)
}

File: orchestration/src/tailoring/openai.ts
import OpenAI from "openai";

const rawModel = process.env.MODEL ?? "gpt-4o-mini";
export const tailorModel = rawModel.includes("/") ? rawModel.split("/").pop()! : rawModel;

export const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

export async function callJsonLLM<T>(
  system: string,
  user: string,
  schema: {
    name: string;
    schema: Record<string, unknown>;
  },
  temperature = 0.25,
): Promise<T> {
  const res = await openai.chat.completions.create({
    model: tailorModel,
    messages: [
      { role: "system", content: system },
      { role: "user", content: user },
    ],
    response_format: {
      type: "json_schema",
      json_schema: {
        name: schema.name,
        strict: true,
        schema: schema.schema,
      },
    },
    temperature,
  });
  const raw = res.choices[0]?.message?.content;
  if (!raw) throw new Error("Empty LLM response");
  return JSON.parse(raw) as T;
}
